File: Concordance_Java/src/concordance/RBTree.java
/**************************************************
 * Red Black Tree:                                *
 * this class implement a RB-Tree together        *
 * with the operations for inserting nodes in it. *
 * RB-Tree is implemented using a TreeMap.        *
 * The key of the elements are the tokens while   *
 * the values stored are the infos about the      *
 * tokens returned by the parser.                 *
 * Operations are performed in O(log n) time      *
 * where n is the number of tokens in the set.    *
 * Filling the tree takes time O(n log n).        *
 * Space is O(n log n).                           *
 * Printing the elements of the tree takes        *
 * O(n) time and the elements are printed         *
 * following the ordering used for insertion.     *
 **************************************************/
package concordance;

import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.TreeMap;

/**
 *
 * @author fedecampe
 */
class RBTree extends AbstractDataStore {
  private String dbg;
  /* map representing the tree */
  private TreeMap< String, ArrayList<Content> > rbTree;
  
  public RBTree () {
    dbg  = "#log: RB_Tree - ";
    rbTree = new TreeMap< String, ArrayList<Content> >();
  }//RBTree

  /* Fill RB-tree using a parser */
  @Override
  void fillStore(Parser prs) {
    if ( verbose ) {
      System.out.println( dbg + "Start filling store...");
    }
    
    if ( timer ) setTime();
    
    /* Invoke parser to get all the
     * contents to add (on-line) into the map. */
    while ( prs.hasMoreElements() ) {
      addContent ( prs.getNextElement() );
    }
    
    if ( timer ) {
      System.out.print( dbg + "Time to fill the tree: ");
      System.out.println( getTime() + "ms" );
    }
    
    if ( verbose ) {
      System.out.println( dbg + "...Filled store.");
    }
  }//fillStore
  
   /* Size of the tree:
    * number of nodes in the tree. */
  @Override
  public int getSize() {
    return rbTree.size();
  }//getSize
  
   /* Add content to the tree.
    * Inset a new string with the info
    * specified in the content.
    * Otherwise update the content if the key is
    * already present. */
  @Override
  public void addContent( Content content ) {
    if ( content != null ) {
      String token = (String) content.getRefBaseObj();
      if ( rbTree.containsKey( token ) ) {
        rbTree.get( token ).add( content );
      }
      else {
        ArrayList<Content> arrAux = new ArrayList<Content>();
        arrAux.add(content);
        rbTree.put( token , arrAux );
      }
    }
  }//addContent

   /*Visit tree (it prints the sorted list of
    * the contents of the nodes). */
  @Override
  public void print() {
    
    if ( verbose ) {
      System.out.println( dbg + "Start visiting RB-Tree");
      System.out.println( dbg + "Visiting RB-Tree...");
      System.out.println( "---------------- CONCORDANCE ------------------");
    }
    
    if ( timer ) setTime();
    
    boolean firstPrint;
    String strToPrint = "";
    for ( String key: rbTree.keySet() ) {
      firstPrint = true;
      ArrayList<Content> currentCnt = rbTree.get( key );
      for ( int i = 0; i < currentCnt.size(); i++ ) {
        if ( firstPrint ) {
          strToPrint += key;
          for ( int ii = 0; ii < (25 - (key.length())); ii++ ) strToPrint += " ";
          strToPrint += "{";
          strToPrint += currentCnt.size();
          strToPrint += ":";
          firstPrint = false;
        }
        
        Content content = currentCnt.get(i);
        ArrayList<Integer> idxLine = ((Token)content).getLines();
        for ( int ii = 0; ii < idxLine.size(); ii++ ) {
          if ( i < currentCnt.size() - 1 ) {
            strToPrint += idxLine.get(ii) + ",";
          } 
          else {
            strToPrint += idxLine.get(ii);
          }
        }//ii 
      }//i
      strToPrint += "}\n";
    }//key
    
    if ( outFile.equals("") ) {
      System.out.append(strToPrint);
    }
    else {
      try {
        BufferedWriter bw = new BufferedWriter( new FileWriter(outFile) );
        bw.write(strToPrint);
        bw.close();
      } 
     catch ( IOException e ) {
        System.out.println( "Can't open the file " + outFile );
      }
    }
    
    if ( verbose ) {
      System.out.println( "-----------------------------------------------" );
    }
    if ( timer ) {
      System.out.print( dbg + "Time to visit RB-Tree: ");
      System.out.println( getTime() + "ms" );
    }
    
  }//print
  
}//RBTree

File: Concordance_C++/host_core/rb_tree.cpp
#include "rb_tree.h"
#include "token.h"

//#define RB_TREE_DBG

using namespace std;

RB_Tree::RB_Tree () {
  _dbg  = "#log: RB_Tree - ";
}//-

RB_Tree::~RB_Tree () {
  _rb_tree.clear();
}//-

int
RB_Tree::get_size() {
  return _rb_tree.size();
}//get_size

void
RB_Tree::add_content ( Content* content ) {
  if ( content != nullptr ) {
#ifdef RB_TREE_DBG
    content->print();
#endif
    
    Token * casted_token = static_cast< Token * > ( const_cast<Content *>(content) );
    string key = *((string *) content->get_ref_base_obj());
    _rb_tree[ key ].push_back( content );
  }
}//add_content

void
RB_Tree::fill_store ( Parser * parser ) {
  if ( _verbose ) {
    std::cout << _dbg << "Start filling store...\n";
  }
  
  if ( _timer ) set_timer();

  /*
   * Invoke parser to get all the
   * contents to add (on-line) into the map.
   */
  while ( parser->has_more_elements() ) {
    add_content ( parser->get_next_content() );
  }
  
  if ( _timer ) {
    std::cout << _dbg << "Time for filling the tree: ";
    std::cout << get_timer() << " sec.\n";
  }
  
  if ( _verbose ) {
    std::cout << _dbg << "...Filled store.\n";
  }
}//fill_store

void
RB_Tree::print () {
  
  if ( _verbose ) {
    std::cout << _dbg << "Start visiting RB-Tree\n";
  }
  
  bool first_print;
  stringstream str_stream;
  string token_string;
  string lines;
  
  if ( _verbose ) {
    std::cout << _dbg << "Visiting RB-Tree...\n";
    std::cout << std::endl;
    std::cout << "---------------- CONCORDANCE ------------------\n";
  }
  
  if ( _timer ) set_timer();
  
  for ( auto mapped_value : _rb_tree ) {
    first_print = true;
    Content * cont;
    int size = mapped_value.second.size();
    for ( int i = 0; i < size; i++ ) {
      cont = (mapped_value.second)[ i ];
    //for ( auto cont : mapped_value.second ) {
      Token * casted_token = static_cast< Token * > ( const_cast<Content *>(cont) );
      
      if ( first_print ) {
        token_string = casted_token->get_string();
        str_stream << token_string;
        for ( int i = 0; i < (25 - (token_string.length())); i++ ) str_stream << " ";
        str_stream << "{";
        str_stream << mapped_value.second.size();
        str_stream << ":";
        first_print = false;
      }
      
      vector <int> lines = casted_token->get_lines();
      for ( int ii = 0; ii < lines.size(); ii++ ) {
        if (i < size - 1) {
          str_stream << lines[ ii ] <<",";
        }
        else {
          str_stream << lines[ ii ];
        }
      }
      lines.clear();
    }
    str_stream << "}\n";
  }
  
  if ( _verbose ) {
    std::cout << "-----------------------------------------------\n";
    std::cout << std::endl;
  }
  
  double tot_time;
  if ( _timer ) tot_time = get_timer ();
  
  string out_str = str_stream.str();
  /// Print the resulting string
  if ( _out_file.compare ( "" ) ) {
    std::ofstream ofs ( _out_file, std::ofstream::out );
    if ( ofs.is_open() ) {
      ofs << out_str;
      ofs.close();
    }
    else {
      cout << "Can't open file: " << _out_file << endl;
    }
  }
  else {
    cout << out_str;
  }
  
  if ( _timer ) {
    std::cout << _dbg << "Time for visiting the RB-Tree: ";
    std::cout << tot_time << " sec.\n";
  }
}//print
